easy_phi/hwconf.py

```diff
@@ -0,0 +1,91 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+
+"""This module provides functions to obtain hardware configuration and keep it up to date.
+
+"""
+
+import pyudev
+import hwal
+from tornado.options import options
+
+callbacks = [
+    # hardware configuration change listener will call these methods
+    # it is necessary to send updates to websockets
+]
+
+# device #0 represents broadcast
+modules = [None for i in range(options.slots + 1)]
+modules[0] = hwal.BroadcastModule(modules)
+
+_context = pyudev.Context()
+
+
+def get_rack_slot(device):
+    global modules  # Ugly, I know. Let me know if there is a better way
+
+    # first, check if device is already represented in modules
+    for i, module in enumerate(modules):
+        if module.device == device:
+            return i
+
+    # if it is not in the list, match it with USB ports of the rack
+    # TODO: get usb ports list from config file (command line option)
+    slot = 1
+
+    # if device is neither in modules list nor port is associated with rack slot, assign to first free slot
+    if not 1 <= slot <= options.slots:
+        for i in range(options.slots + 1, len(modules)):
+            if modules[i] is None:
+                return i
```

Comment:
I'll try to. 
rack_slot.conf
>Slot 1 <--> USB Port 1
>Slot 2 <--> USB Port 2

Step 1. New module is connected to USB Port 1 -> module is assigned to Rack Slot 1
Step 2. New module is connected to USB Port 3 -> port is not associated with any rack slot -> module is assigned to the first free slot - Rack Slot 2 
Step 3. New module is connected to USB Port 2 -> module is assigned to Rack Slot 2
Result: module from Step 2 is lost from the array